### PR TITLE
[AArch64][NEON] Add intrinsics for LUTI

### PR DESCRIPTION
**Pull request by Lukacma**
This patch adds intrinsics for NEON LUTI2 and LUTI4 instructions as specified in the [ACLE proposal](https://github.com/ARM-software/acle/pull/324)

### CODE REVIEW

File: clang/include/clang/Basic/arm_neon.td [hunk 1]
```diff
@@ -2096,3 +2096,19 @@ let ArchGuard = "defined(__aarch64__) || defined(__arm64ec__)", TargetGuard = "r
   def VLDAP1_LANE : WInst<"vldap1_lane", ".(c*!).I", "QUlQlUlldQdPlQPl">;
   def VSTL1_LANE  : WInst<"vstl1_lane", "v*(.!)I", "QUlQlUlldQdPlQPl">;
 }
+
+//Lookup table read with 2-bit/4-bit indices
```
Comment (davemgreen) [inline clang/include/clang/Basic/arm_neon.td:2100]: // Lookup

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 2]
```diff
@@ -6420,6 +6420,76 @@ def : Pat<(v16i8 (int_aarch64_neon_tbx1 (v16i8 V128:$Rd),
...
+                    (v8i16 VecListOne8h:$Rn2), (v16i8 V128:$Rm), 
+                    (i32 VectorIndexS32b_timm:$idx))),
+            (LUT4_H (REG_SEQUENCE QQ, VecListOne8h:$Rn1, qsub0, VecListOne8h:$Rn2, qsub1), V128:$Rm,  VectorIndexS32b_timm:$idx)>;
+  def : Pat<(v8f16 (int_aarch64_neon_vluti4q_laneq_x2 (v8f16 VecListOne8h:$Rn1), 
+                    (v8f16 VecListOne8h:$Rn2), (v16i8 V128:$Rm), 
+                    (i32 VectorIndexS32b_timm:$idx))),
+            (LUT4_H (REG_SEQUENCE QQ, VecListOne8h:$Rn1, qsub0, VecListOne8h:$Rn2, qsub1), V128:$Rm,  VectorIndexS32b_timm:$idx)>;
+}
+
+let Predicates = [HasLUT, HasBF16] in {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:6475]: I think you can make this HasLUT only without needing HasBF16, like the fp16 versions above. Unless that doesn't work? It should only really be dependent on the size of the register (and HasLUT, obviously).

You might be able to make a multiclass too for the Pats with a parameter for the type, if they could shares a lot of the same code.

**Comment by efriedma-quic** [comment]
To clarify for anyone else looking at this... there are three families of LUTI instructions: one uses NEON registers, one uses SVE registers, and one uses SME registers.  This patch is just the variant that uses NEON registers.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 2 above)
Comment (Lukacma) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Thank you for the suggestion. I have , hopefully, cleaned up patterns now.

File: clang/include/clang/Basic/arm_neon.td (hunk 1 above)
Comment (Lukacma) [reply clang/include/clang/Basic/arm_neon.td]: Done

**Review by davemgreen — COMMENTED** [review]
Thanks this looks great. I've not checked the C / ACLE intrinsics though - I will defer to @CarolineConcatto and @momchil-velikov for those parts if that is OK.

File: clang/test/CodeGen/aarch64-neon-luti.c [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+// CHECK-LABEL: define dso_local <8 x i16> @test_vluti2_lane_u16(
+// CHECK-SAME: <4 x i16> noundef [[VN:%.*]], <8 x i8> noundef [[VM:%.*]]) local_unnamed_addr #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[TMP0:%.*]] = bitcast <4 x i16> [[VN]] to <8 x i8>
+// CHECK-NEXT:    [[VLUTI2_LANE:%.*]] = tail call <8 x i16> @llvm.aarch64.neon.vluti2.lane.v8i16.v8i8.v8i8(<8 x i8> [[TMP0]], <8 x i8> [[VM]], i32 0)
+// CHECK-NEXT:    ret <8 x i16> [[VLUTI2_LANE]]
+//
+uint16x8_t test_vluti2_lane_u16(uint16x4_t vn, uint8x8_t vm) {
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-neon-luti.c:132]: This test is failing when trying to lower:
fatal error: error in backend: Cannot select: intrinsic %llvm.aarch64.neon.vluti2.lane
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: ./bin/clang -target aarch64-linux-gnu -march=armv8-a+lut+bf16+simd -S -o - ../clang/test/CodeGen/aarch64-neon-luti.c
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module '../clang/test/CodeGen/aarch64-neon-luti.c'.
4.	Running pass 'AArch64 Instruction Selection' on function '@test_vluti2_lane_u16'

File: clang/test/CodeGen/aarch64-neon-luti.c [hunk 4]
```diff
@@ -0,0 +1,433 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+#include <arm_neon.h>
+// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +neon -target-feature +lut -target-feature +bf16 -O3 -emit-llvm -o - %s | FileCheck %s
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-neon-luti.c:3]: We tend to add this RUN line to check if all the tests can lower:
// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +neon -target-feature +lut -target-feature +bf16 -S -disable-O0-optnone -Werror -Wall -o /dev/null %s

File: llvm/include/llvm/IR/IntrinsicsAArch64.td [hunk 5]
```diff
@@ -545,6 +545,25 @@ let TargetPrefix = "aarch64", IntrProperties = [IntrNoMem] in {
   def int_aarch64_neon_vcmla_rot270 : AdvSIMD_3VectorArg_Intrinsic;
 }
 
+let TargetPrefix = "aarch64" in {
+def int_aarch64_neon_vluti2_lane : DefaultAttrsIntrinsic<[llvm_anyvector_ty],
```
[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/include/llvm/IR/IntrinsicsAArch64.td:549]: I don't have strong opinion either way. I do not see advantage to either approach, but maybe I am missing smth

[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/include/llvm/IR/IntrinsicsAArch64.td:549]: In a sense. The indices are integral type, but they are polymorphic in vector size.

[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/include/llvm/IR/IntrinsicsAArch64.td:549]: Done

File: clang/test/CodeGen/aarch64-neon-luti.c (hunk 4 above)
Comment (Lukacma) [reply clang/test/CodeGen/aarch64-neon-luti.c:3]: Done

File: clang/test/CodeGen/aarch64-neon-luti.c (hunk 3 above)
Comment (Lukacma) [reply clang/test/CodeGen/aarch64-neon-luti.c:132]: Fixed

**Review by CarolineConcatto — APPROVED** [review]

**Pull request merged**